File: erlmonads_src/examples.hs
#!/usr/bin/env runhaskell

import Control.Monad.State
import Data.List

{-
Определение State-монады на основе трансформера
https://github.com/ghc/packages-transformers/blob/master/Control/Monad/Trans/State/Strict.hs#L85


https://github.com/ghc/ghc/blob/master/libraries/base/Control/Monad.hs
-}

-- Example 1 (factorial)

fact_rec :: Int -> State Int Int
fact_rec 0 = do
    acc <- get
    return acc
fact_rec n = do
    acc <- get
    put (acc * n)
    fact_rec (n - 1)

fact :: Int -> Int
fact n = evalState (fact_rec n) 1
-- fact n = fst $ runState (fact_rec n) 1

example1 = do
    putStrLn "Example1 (factorial):"
    print (fact 3)


-- Example 2 (fibonacci)

fib_step :: Int -> State (Int, Int) ()
fib_step n = do 
    (a,b) <- get
    put (b,a+b)

fib_v1 :: Int -> Int
fib_v1 n = flip evalState (0,1) $ do 
    forM [0..(n-1)] $ fib_step
    (a,b) <- get
    return a

fib_m_step :: State (Int, Int) ()
fib_m_step = 
    modify $ \(a,b) -> (b,a+b)

fib_v2 :: Int -> Int
fib_v2 n = flip evalState (0,1) $ do
    replicateM n fib_m_step
    gets fst

fib_v3 :: Int -> Int
fib_v3 n = flip evalState (0,1) $ do
    sequence (Data.List.replicate n fib_m_step)
    gets fst

example2 = do
    putStrLn "Example2 (fibonacci):"
    print (fib_v1 4)
    print (fib_v2 4)
    print (fib_v3 4)



main = do
    example1
    example2

File: erlmonads_src/examples.erl
#!/usr/bin/env escript
%% -*- erlang -*-
%%! -pz ./erlmonads/ebin

-compile({parse_transform, erlmonads_expr_do}).

%code:add_pathz("./erlmonads/ebin").
%code:add_pathz("./erlmonads/src/").

% case_one() ->
%     io:format("Case 1:~n"),
%     Data = {[
%         {<<"key1">>, <<"value1">>},
%         {<<"key2">>, {[
%             {<<"key2.1">>, <<"value2.1">>},
%             {<<"key2.2">>, <<"value2.2">>}
%         ]}},
%         {<<"key3">>, <<"value3">>}
%     ]},
%     V21 = ej:get([<<"key2">>, <<"key2.1">>], Data),
%     V21upd = <<"Updated1 ", V21/binary>>,

%     V22 = ej:get([<<"key2">>, <<"key2.2">>], Data),
%     V22upd = <<"Updated2 ", V22/binary>>,

%     DataUpd = ej:set([<<"key2">>, <<"key2.1">>],
%             ej:set([<<"key2">>, <<"key2.2">>], Data, V22upd),
%         V21upd),

%     % --------------

%     Upd1 = fun(Data) -> 
%         V = ej:get([<<"key2">>, <<"key2.1">>], Data),
%         Vupd = <<"Updated1 ", V/binary>>,
%         ej:set([<<"key2">>, <<"key2.1">>], Data, Vupd)
%     end,

%     Upd2 = fun(Data) -> 
%         V = ej:get([<<"key2">>, <<"key2.2">>], Data),
%         Vupd = <<"Updated2 ", V/binary>>,
%         ej:set([<<"key2">>, <<"key2.2">>], Data, Vupd)
%     end,

%     io:format("~p~n", [DataUpd]),
%     io:format("~p~n", [Upd1(Upd2(Data))]),
%     ok.

% case_one_with_monad() ->
%     io:format("Case 1 (with monad):~n"),
%     State = erlmonads_state_t:new(erlmonads_identity),
%     ok.


%% Example 1

% fact_rec :: Int -> State Int Int
% fact_rec 0 = do
%     acc <- get
%     return acc
% fact_rec n = do
%     acc <- get
%     put (acc * n)
%     fact_rec (n - 1)

% fact :: Int -> Int
% fact n = fst $ runState (fact_rec n) 1

fib_m_step(State) ->
    State:modify(fun ({X, Y}) -> {Y, X+Y} end).

fib_v1(N) ->
    State = erlmonads_state_t:new(erlmonads_identity),
    {_, R} = State:exec(
        erlmonads:sequence(
            State,
            lists:duplicate(N, fib_m_step(State))
        ),
        {0, 1}
    ),
    R.

example1() ->
    io:format("Example 1~n"),
    N = fib_v1(3),
    io:format("~w~n",[N]).


main(_) ->
    example1(),
    ok.